install-deps.sh

```diff
@@ -105,6 +105,29 @@ ENDOFKEY
     $SUDO ln -nsf /usr/bin/g++ /usr/bin/$(uname -m)-linux-gnu-g++
 }
 
+function install_pkg_on_ubuntu {
+    local project=$1
+    shift
+    local sha1=$1
+    shift
+    local codename=$1
+    shift
+    local pkgs=$@
+    local missing_pkgs
+    for pkg in $pkgs; do
+	if ! dpkg -s $pkg &> /dev/null; then
+	    missing_pkgs+=" $pkg"
+	fi
+    done
+    if test -n "$missing_pkgs"; then
+	$SUDO tee /etc/apt/sources.list.d/$project.list <<EOF
+deb https://chacra.ceph.com/r/$project/master/$sha1/ubuntu/$codename/flavors/default $codename main
```

Comment:
This line also needs the unauthenticated settings so that `apt-get update` doesn't break when installing other packages.

    deb [trusted=yes] https://chacra.ceph.com/r/$project/master/$sha1/ubuntu/$codename/flavors/default $codename main

Alternatively, you could also allow the "latest" to be used (instead of requiring a sha1), and finally, if this repo reports to shaman, this could be further automated by asking shaman to point to the repo file. For example, try this:

    curl https://shaman.ceph.com/api/repos/ceph/luminous/latest/ubuntu/xenial/repo

Reply:
> are you always going to default to master? This is fine, just making sure this is really what you need here
 
i thought we could always map the "master" in repo to "master" branch. but turns out we have also the "ref" (i.e. `$sha1` here). and it could annoying if we aways need to update `install-deps.sh` when cutting off a new branch.

Reply:
> you could also allow the "latest" to be used (instead of requiring a sha1)

very cool! but this prevents us from having a repeatable build.

Reply:
> Is it fine to include this line in /etc/apt/sources.list.d/$project.list if it exists already?

good question. i think we can just blindly `cat` this line into this file, and blindly install the dependency packages. is this what you are proposing?

Reply:
Seems safe to do `cat` here. Sounds good!